prototypes/matrix_free_non_linear_poisson/matrix_free_non_linear_poisson.cc

```diff
@@ -0,0 +1,1221 @@
+  const std::pair<unsigned int, unsigned int> &     cell_range) const
+{
+  FEEvaluation<dim, fe_degree, fe_degree + 1, 1, double> phi(data);
+  SourceTerm<dim>                                        source_term_function;
+
+  for (unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+    {
+      phi.reinit(cell);
+
+      phi.read_dof_values_plain(src);
+      phi.evaluate(EvaluationFlags::values | EvaluationFlags::gradients);
+
+      for (unsigned int q = 0; q < phi.n_q_points; ++q)
+        {
+          VectorizedArray<double> source_value = VectorizedArray<double>(0.0);
+
+          if (parameters.source_term == Settings::mms)
+            {
+              Point<dim, VectorizedArray<double>> point_batch =
+                phi.quadrature_point(q);
+
+              for (unsigned int v = 0; v < VectorizedArray<double>::size(); ++v)
+                {
+                  Point<dim> single_point;
+                  for (unsigned int d = 0; d < dim; ++d)
+                    single_point[d] = point_batch[d][v];
+                  source_value[v] = source_term_function.value(single_point);
+                }
+            }
+          phi.submit_value(-std::exp(phi.get_value(q)) - source_value, q);
+          phi.submit_gradient(phi.get_gradient(q), q);
+        }
+
+      phi.integrate_scatter(EvaluationFlags::values |
+                              EvaluationFlags::gradients,
+                            dst);
+    }
+}
+
+template <int dim, int fe_degree>
+void
+MatrixFreePoissonProblem<dim, fe_degree>::assemble_rhs()
+{
+  TimerOutput::Scope t(computing_timer, "assemble right hand side");
+
+  evaluate_residual(system_rhs, solution);
+
+  system_rhs *= -1.0;
+}
+
+template <int dim, int fe_degree>
+double
+MatrixFreePoissonProblem<dim, fe_degree>::compute_residual(const double alpha)
+{
+  TimerOutput::Scope t(computing_timer, "compute residual");
+
+  LinearAlgebra::distributed::Vector<double> residual;
+  LinearAlgebra::distributed::Vector<double> evaluation_point;
+
+  system_matrix.initialize_dof_vector(residual);
+  system_matrix.initialize_dof_vector(evaluation_point);
+
+  evaluation_point = solution;
+  if (alpha > 1e-12)
+    {
+      evaluation_point.add(alpha, newton_update);
+    }
+
+  evaluate_residual(residual, evaluation_point);
+
+  return residual.l2_norm();
+}
+
+template <int dim, int fe_degree>
+void
+MatrixFreePoissonProblem<dim, fe_degree>::compute_update()
+{
+  TimerOutput::Scope t(computing_timer, "compute update");
+
+  solution.update_ghost_values();
+
+  system_matrix.evaluate_newton_step(solution);
+
+  mg_transfer.interpolate_to_mg(dof_handler, mg_solution, solution);
+
+  using SmootherType =
+    PreconditionChebyshev<LevelMatrixType,
+                          LinearAlgebra::distributed::Vector<float>>;
+  mg::SmootherRelaxation<SmootherType,
+                         LinearAlgebra::distributed::Vector<float>>
+                                                       mg_smoother;
```

Comment:
Why use a different data type here?

Reply:
This is just a way of exploiting single precision. It has been shown that for these kinds of problems one can have a mixed-precision approach, where one can use double precision for the CG and single-precision for the multigrid as preconditioner. This preserves the accuracy and iteration number of a full double precision and it allows you take advantage of vectorization. I used this idea only for the Poisson prototype, the advection-diffusion one uses the same type for both the solver and the preconditioner...

Reply:
Interesting. Clearly this gives an acceleration both from a memory and FLOPS perspective. Cool stuff.